csharp/rust/src/lib.rs

```diff
@@ -1,284 +1,472 @@
-        } else {
-            Some(
-                CStr::from_ptr(file_name)
-                    .to_str()
-                    .expect("Can not read string argument."),
-            )
+pub extern "C-unwind" fn csharp_command(
+    in_client_ptr: *const c_void,
+    in_callback: CommandCallback,
+    in_callback_data: *mut c_void,
+    in_request_type: RequestType,
+    in_routing_info: *const routing::RoutingInfo,
+    // ToDo: Rework into parameter struct (understand how Command.arg(...) works first)
+    //       handling the different input types.
+    in_args: *const *const c_char,
+    in_args_count: c_int,
+    // ToDo: Pass in ActivityContext and connect C# OTEL with Rust OTEL
+) -> CommandResult {
+    logger_core::log(Trace, "csharp_ffi", "Entered csharp_command");
+    if in_client_ptr.is_null() {
+        logger_core::log(Error, 
+            "csharp_ffi",
+            "Error in csharp_command called with null handle",
+        );
+        return CommandResult::new_error(helpers::to_cstr_ptr_or_null("Null handle passed"));
+    }
+    let args = match helpers::grab_vec_str(in_args, in_args_count as usize) {
+        Ok(d) => d,
+        Err(e) => {
+            logger_core::log(Error, 
+                "csharp_ffi",
+                format!("Error in string transformation: {:?}", e.to_string()),
+            );
+            return match e {
+                Utf8OrEmptyError::Utf8Error(e) => {
+                    CommandResult::new_error(helpers::to_cstr_ptr_or_null(e.to_string().as_str()))
+                }
+                Utf8OrEmptyError::Empty => CommandResult::new_error(helpers::to_cstr_ptr_or_null(
+                    "Null value passed for host",
+                )),
+            };
+        }
+    };
+    let cmd = match in_request_type.get_command() {
+        None => {
+            logger_core::log(Error, 
+                "csharp_ffi",
+                "Error in csharp_command called with unknown request type",
+            );
+            return CommandResult::new_error(helpers::to_cstr_ptr_or_null("Unknown request type"));
+        }
+        Some(d) => d,
+    };
+    let callback = in_callback;
+    let callback_data = in_callback_data as usize;
+
+    let ffi_handle = unsafe { Box::leak(Box::from_raw(in_client_ptr as *mut FFIHandle)) };
+    let handle = ffi_handle.handle.clone();
+    let routing_info = if in_routing_info.is_null() {
+        None
+    } else {
+        Some(unsafe {
+            match (*in_routing_info).to_redis() {
+                Ok(d) => d,
+                Err(e) => {
+                    logger_core::log(Error, 
+                        "csharp_ffi",
+                        format!(
+                            "Error while parsing route in string transformation: {:?}",
+                            e.to_string()
+                        ),
+                    );
+                    return match e {
+                        Utf8OrEmptyError::Utf8Error(e) => CommandResult::new_error(
```

Comment:
Dislike commands, routes support only simple strings. We can pass them as `CStr` or as `byte[]`, but you'll never have UTF8Error. If something went wrong with route - just return `RequestError`.

Reply:
If a user is using the library as-is with their own bindings, UTF errors may occur. This accounts for the possibility that the calling side could be anything, including manually written wrappers around Glide.